### PR TITLE
Visualisation: Allow different shapes agent portrayals#2164 

### PR DESCRIPTION
**Pull request by Aryan1Mahahjan**
Hi, this is Draft Solution to #2164.

I was thinking about using the below, which use the Encode function attribute 

The added code snippet checks for shape and fill keys in agent data. If present, it adds them to the encoding_dict using the corresponding keys shape and fill. This allows you to define visual properties like circle shapes and filled elements for agents in your visualization.

Draft: Addition of 2 keys in enconding_dict: shape(str) and fill(bool)

### CODE REVIEW

**Comment by Aryan1Mahahjan** [comment]
Commit Issue with pytest:

I'm having trouble committing changes. When I try to commit, I see an error message related to pytest. Here's the output:

image

**Comment by EwoutH** [comment]
Thanks for the PR!

Weird, seems totally unrelated to these changes or our CI. How are you committing?

**Comment by rht** [comment]
Regarding with the typos, we need to add `.venv` to .gitignore, because it is a common name for a local virtualenv folder.
Regarding with the PR, it would be great to have a screenshot showcasing that it works in action, by modifying the Schelling_experimental example.

**Comment by Aryan1Mahahjan** [comment]
I reset my project to an earlier state using Git.

 Now, I'm trying to re-install required dependencies like ruff and pytest, but I'm unable to activate the virtual environment (.venv). This prevents me from installing the packages currently.

**Comment by rht** [comment]
> Now, I'm trying to re-install required dependencies like ruff and pytest, but I'm unable to activate the virtual environment (.venv). This prevents me from installing the packages currently.

Can you share the error messages?

**Comment by Aryan1Mahahjan** [comment]
image
I did not have this issue when installing the test last time.

**Comment by rht** [comment]
Your .venv is probably gone (the other possibility is that .venv exists but .venv/bin/activate is not executable, but doesn't have to be executable anyway). You likely have to recreate the virtualenv from scratch.

**Comment by EwoutH** [comment]
Maybe it's easiest to reinstall at this point.

File: mesa/visualization/components/altair.py [hunk 1]
```diff
@@ -53,6 +53,12 @@ def portray(g):
     has_size = "size" in all_agent_data[0]
     if has_size:
         encoding_dict["size"] = alt.Size("size", type="quantitative")
+    has_shape = "shape" in all_agent_data[0]
+    if has_shape:
+        encoding_dict["shape"] = agent_data["shape"]
+    has_fill = "fill" in all_agent_data[0]
+    if has_shape:
```
Comment (EwoutH) [inline mesa/visualization/components/altair.py:60]: I thin this should be
```suggestion
    if has_fill:
```

**Comment by Aryan1Mahahjan** [comment]
So, Majority of the tests are passing.
image
There is a 1 failed test and 1 Error.
image

**Comment by EwoutH** [comment]
In CI all the tests are passing, so and the warnings are unrelated to this PR. So test-wise we’re good!

Could you update the title of the PR and add a screenshot of how it now looks to the first PR message?

**Comment by Aryan1Mahahjan** [comment]
image

image

**Comment by EwoutH** [comment]
Sorry, I mean a screenshot of how the visualisation looks with different shapes, to show that it works.

**Comment by Aryan1Mahahjan** [comment]
> Sorry, I mean a screenshot of how the visualisation looks with different shapes, to show that it works.

How should show or generate visualization looks with different is their a test or dataset to show?

**Comment by EwoutH** [comment]
You could adapt an example model, like: https://github.com/projectmesa/mesa-examples/tree/main/examples/boltzmann_wealth_model_experimental

Or create a simple model yourself.

**Comment by EwoutH** [comment]
@Aryan1Mahahjan would you still like to work on this?

**Comment by EwoutH** [comment]
This now has been solved in #2214.

**Pull request closed without merging**